### ISSUE TITLE
Question: (active) training loop

### ISSUE THREAD

**Issue by pietrolesci** [BlackHC/batchbald_redux#1]
Hi @BlackHC,

Thanks a lot for creating this amazing library. I have a question regarding the (active) training loop in the example [here](https://github.com/BlackHC/batchbald_redux/blob/master/06_example_experiment.ipynb):

At the end of each iteration in the while loop we acquire the new data via

```python
active_learning_data.acquire(candidate_batch.indices)
```

That is, the `active_learning_data.training_dataset` will have new instances. Here it comes my doubt, 

> how the acquired instances are then included in the `train_loader`? 

The `train_loader` is defined outside the while loop and does not change throughout training. It seems to me that the model sees always the same data in the training loop

```python
# Train
for data, target in tqdm(train_loader, desc="Training", leave=False):
    data = data.to(device=device)
    target = target.to(device=device)
     ...
```

> Shouldn't it be able to train on the "new" (extended) training set, i.e. shouldn't the train_loader be redefined at each iteration?

To provide a bit of context: I am trying to write a model leveraging both `batchbald_redux` and `baal`. Out of curiosity, can you comment on the differences between the two libraries and why you chose to reimplement some things (e.g., the ActiveLearningData?). I am trying to combine them and get the best of each one. Also, I have seen that they included an implementation of BatchBald taken from the code you provided originally with the paper, is there an advantage on using that implementation vs this one? - Well, I guess these are a bonus questions, feel free to ignore :)

Thanks a lot in advance for your attention.

Best,
Pietro

**Comment by BlackHC** [comment]
Dear Pietro,

thank you for your question!

The `train_loader` is using the dataset `active_learning_data.training_dataset`. It is not immutable but changes as new samples are acquired. The overlying dataloader will thus iterate over the updated dataset.

I wanted for `batchbald_redux` to be independent of `baal`. It is based on the original BatchBALD code, which might or might not have served as an inspiration to `baal`. 

In more practical terms, `batchbald_redux` uses `toma` which allows for automatic adaptation of GPU memory usage which potentially might allow the code to run where `baal`'s implementation might not. (I haven't looked at that implementation in a while.)

I hope this answers your question somewhat.

Best wishes,
 Andreas

**Comment by pietrolesci** [comment]
Hi Andreas,

I really appreciated your reply. I have been experimenting with it lately and found that it fact it does change without reloading it. Thanks to your comment now I understand why :)

Re: `batchbald_redux` - `baal` relationship, I get it and I think that was a good choice as in this way the library remains self-contained. I am currently using `batchbald_redux` on a project and finding it amazing (+ toma is the real cherry on the cake). I was wondering about the best way to reach out to you in case I have comments/questions. Would Github issues work for this?

Once again, thank you very much for your consideration.

Best,
Pietro

**Comment by BlackHC** [comment]
Dear Pietro,

thanks and sorry for the late reply. I'm happy you like the library.

GitHub is one way to reach me. You can also message me on Twitter.

Best wishes,
 Andreas
